Mathlib/Data/BitVec/Lemmas.lean

```diff
@@ -1,199 +0,0 @@
-
-/-!
-### Distributivity of `BitVec.ofFin`
--/
-section
-variable (x y : Fin (2^w))
-
-@[simp] lemma ofFin_neg : ofFin (-x) = -(ofFin x) := by
-  ext; rw [neg_eq_zero_sub]; simp; rfl -- v4.7.0-rc1 issues remove?
-
-@[simp] lemma ofFin_and : ofFin (x &&& y) = ofFin x &&& ofFin y := by
-  ext
-  simp [HAnd.hAnd, AndOp.and, Fin.land, BitVec.and, toNat_ofFin]
-  -- v4.7.0-rc1 issues remove?
-
-@[simp] lemma ofFin_or  : ofFin (x ||| y) = ofFin x ||| ofFin y := by
-  ext
-  simp [HOr.hOr, OrOp.or, Fin.lor, BitVec.or, toNat_ofFin]
-  -- v4.7.0-rc1 issues remove?
-
-@[simp] lemma ofFin_xor : ofFin (x ^^^ y) = ofFin x ^^^ ofFin y := by
-  ext
-  simp [HXor.hXor, Xor.xor, Fin.xor, BitVec.xor, toNat_ofFin]
-  -- v4.7.0-rc1 issues remove?
-
-@[simp] lemma ofFin_mul : ofFin (x * y)   = ofFin x * ofFin y   := rfl
-
--- These should be simp, but Std's simp-lemmas do not allow this yet.
-lemma ofFin_zero : ofFin (0 : Fin (2^w)) = 0 := rfl
-lemma ofFin_one  : ofFin (1 : Fin (2^w)) = 1 := by
-  simp only [OfNat.ofNat, BitVec.ofNat, and_pow_two_is_mod]
-
-lemma ofFin_nsmul (n : ℕ) (x : Fin (2^w)) : ofFin (n • x) = n • ofFin x := rfl
-lemma ofFin_zsmul (z : ℤ) (x : Fin (2^w)) : ofFin (z • x) = z • ofFin x := rfl
-@[simp] lemma ofFin_pow (n : ℕ) : ofFin (x ^ n) = ofFin x ^ n := rfl
-
-@[simp] lemma ofFin_natCast (n : ℕ) : ofFin (n : Fin (2^w)) = n := by
-  simp only [Nat.cast, NatCast.natCast, OfNat.ofNat, BitVec.ofNat, and_pow_two_is_mod]
-  rfl
-
--- See Note [no_index around OfNat.ofNat]
-@[simp] lemma ofFin_ofNat (n : ℕ) :
-    ofFin (no_index (OfNat.ofNat n : Fin (2^w))) = OfNat.ofNat n := by
-  simp only [OfNat.ofNat, Fin.ofNat', BitVec.ofNat, and_pow_two_is_mod]
-
-end
-
-/-!
-### Distributivity of `Std.BitVec.toFin`
--/
-section
-variable (x y : BitVec w)
-
-@[simp] lemma toFin_neg : toFin (-x) = -(toFin x) := by
-  ext; rw [neg_eq_zero_sub]; simp; rfl
-  -- v4.7.0-rc1 issues remove?
-
--- These should be simp, but Std's simp-lemmas do not allow this yet.
-lemma toFin_zero : toFin (0 : BitVec w) = 0 := rfl
-lemma toFin_one  : toFin (1 : BitVec w) = 1 := by
-  rw [toFin_inj]; simp only [ofNat_eq_ofNat, ofFin_ofNat]
-
-lemma toFin_nsmul (n : ℕ) (x : BitVec w) : toFin (n • x) = n • x.toFin := rfl
-lemma toFin_zsmul (z : ℤ) (x : BitVec w) : toFin (z • x) = z • x.toFin := rfl
-@[simp] lemma toFin_pow (n : ℕ) : toFin (x ^ n) = x.toFin ^ n := rfl
-
-lemma toFin_natCast (n : ℕ) : toFin (n : BitVec w) = n := by
-  rw [toFin_inj]; simp only [ofFin_natCast]
-
-end
-
-/-!
-### `IntCast`
--/
-
--- Either of these follows trivially from the other. Which one to
--- prove is not yet clear.
-proof_wanted ofFin_intCast (z : ℤ) : ofFin (z : Fin (2^w)) = z
-
-proof_wanted toFin_intCast (z : ℤ) : toFin (z : BitVec w) = z
-
-/-!
-## Ring
--/
-
--- TODO: generalize to `CommRing` after `ofFin_intCast` is proven
-instance : CommSemiring (BitVec w) :=
-  toFin_injective.commSemiring _ toFin_zero toFin_one toFin_add toFin_mul toFin_nsmul
-    toFin_pow toFin_natCast
```

Comment:
While most of these files are indeed results that should be in core after the move, this particular instance is about `CommSemiring`, a Mathlib typeclass, and thus should remain in Mathlib.

Reply:
Moreover, we in fact have a [`CommRing` instance](https://github.com/opencompl/ssa/blob/ec6e647cd91936509e372d1615e95aab7415fe8c/SSA/Projects/InstCombine/ForMathlib.lean#L51-L54) which we use for [proof automation](https://github.com/opencompl/ssa/blob/ec6e647cd91936509e372d1615e95aab7415fe8c/SSA/Projects/InstCombine/TacticAuto.lean#L70-L83) that we would eventually like to re-home into `mathlib`.

Reply:
In fact, most proofs that this instance depends on are just def-eqs, so we can inline them if we really don't want those results in mathlib.

Here's a relatively minimal file that retains the `CommSemiring` instance:
```lean4
/-
Copyright (c) 2020 Simon Hudon. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Simon Hudon, Harun Khan, Alex Keizer
-/
import Mathlib.Algebra.Ring.InjSurj
import Mathlib.Data.ZMod.Defs

/-!
# Basic Theorems About Bitvectors

This file contains theorems about bitvectors.
-/

namespace BitVec

variable {w v : Nat}

theorem toFin_injective {n : Nat} : Function.Injective (toFin : BitVec n → _)
  | ⟨_, _⟩, ⟨_, _⟩, rfl => rfl

/-!
## Ring
-/

instance : SMul ℕ (BitVec w) := ⟨fun x y => ofFin <| x • y.toFin⟩
instance : SMul ℤ (BitVec w) := ⟨fun x y => ofFin <| x • y.toFin⟩
instance : Pow (BitVec w) ℕ  := ⟨fun x n => ofFin <| x.toFin ^ n⟩

instance : CommSemiring (BitVec w) :=
  toFin_injective.commSemiring _ rfl rfl toFin_add toFin_mul (fun _ _ => rfl)
    (fun _ _ => rfl) (fun _ => rfl)

end BitVec
```

Reply:
@alexkeizer, do you want to just push that to this PR?

Reply:
> In fact, most proofs that this instance depends on are just def-eqs, so we can inline them if we really don't want those results in mathlib.

Dropping the defeq lemmas seems like a bad idea to me. If mathlib is providing the definitions (`Pow`/`SMul` etc), then it should provide the defeq lemmas as well.

The defeq lemmas about things like `^^^` which are themselves defined upstream should probably migrate upstream.

Reply:
I've now pushed a version of the file I commented which keeps `toFin_nsmul`, `toFin_zsmul` and `toFin_pow`, plus the results about `Function.Injective`

Reply:
I've merged https://github.com/leanprover/lean4/pull/4298

Reply:
Thanks for the push.